### DISCUSSION TITLE
Multi-monitor: Qtile may provide better flexibility to set screen on monitors

### DISCUSSION THREAD

**Discussion by Crystal4276** [qtile/qtile#3839] [Ideas]
From m-col:
"Just in case, because I don't see it mentioned already - note that a monitor can be marked as the 'primary' monitor and this will always be used for the first Screen by Qtile."
https://github.com/qtile/qtile/issues/3186#issuecomment-1013903596

Using Qtile, this behavior has unfortunately consequence regarding the refresh rate of monitors in a X11 multi-monitor configuration.
I was advised by the Arch forum to open an issue/feature request here. 
Here's the full discussion.
https://bbs.archlinux.org/viewtopic.php?id=279714

Quick summary of the issue.
When 60 Hz monitors are combined with a 100 Hz monitor, the only way to get a refresh rate of 100 Hz is to declare the 100Hz monitor as primary monitor in xorg.conf.  The on-screen display properly see a 100 Hz signal coming out of the Nvidia GPU irrespective of the primary monitor status as long as the metamode is properly defined.  Probably due to some client/compositor syncing issues, the "real" refresh rate will be 60 Hz if the 100Hz monitor is not set as primary. Doing so the primary monitor will always be listed as first monitor by xorg and qtile will select it as its first screen.  

It's either:
- every monitor running at 60 Hz with correct screen order in qtile (no primary monitor)
- or main monitor running at 100 Hz(as primary monitor) but then screen order in qtile is impossible to set 

Would it be possible to assign qtile screens to monitors by declaring  xrandr ports like HDMI-0, DP-2 etc... ?

**Comment by clotodex** [comment]
Do you know if this would also be an issue in wayland? @Crystal4276

**Comment by Crystal4276** [comment]
My attempt to solve this issue with wayland was cut short by the fact that i'm using an nvidia card.
So i don't know.

At the end i just I adapted for X11 the scripts i was using to manipulate windows across screen.
For instance
```python
# Send a window within a group to group displayed on left or right screen. 
# Three monitors configuration, screen 0 is on the central monitor (primary) -> Monitors/screen layout: [1,0,2] 
def window_to_previous_screen(qtile, switch_group=False, switch_screen=False):
    i = qtile.screens.index(qtile.current_screen)
    if i == 0:
        group = qtile.screens[i + 1].group.name
        qtile.current_window.togroup(group, switch_group=switch_group)
        if switch_screen == True:
            qtile.to_screen(i + 1)
    if i == 2:
        group = qtile.screens[i - 2].group.name
        qtile.current_window.togroup(group, switch_group=switch_group)
        if switch_screen == True:
            qtile.to_screen(i - 2)

def window_to_next_screen(qtile, switch_group=False, switch_screen=False):
    i = qtile.screens.index(qtile.current_screen)
    if i == 1:
        group = qtile.screens[i - 1].group.name
        qtile.current_window.togroup(group, switch_group=switch_group)
        if switch_screen == True:
            qtile.to_screen(i - 1)
    if i == 0:
        group = qtile.screens[i + 2].group.name
        qtile.current_window.togroup(group, switch_group=switch_group)
        if switch_screen == True:
            qtile.to_screen(i + 2)
```
and 
```python
def _scroll_screen(direction: int) -> callable:
    """
    Scroll to the next/prev group of the subset allocated to a specific screen. This
    will rotate between e.g. 1->2->3->1 when the first screen is focused.
    """
    def _inner(qtile: qtile):
        if len(qtile.screens) == 1:
            current = qtile.groups.index(qtile.current_group)
            destination = (current + direction) % 9
            qtile.groups[destination].toscreen()
            return
        current = qtile.groups.index(qtile.current_group)
        if current < 3:
            destination = (current + direction) % 3
        elif current < 6:
            destination = ((current - 3 + direction) % 3) + 3
        else:
            destination = ((current - 6 + direction) % 3) + 6
        qtile.groups[destination].toscreen()
    return _inner
```
